Junit reporter does not escape error messages and stacktraces

### Describe the bug

The Junit reporter produces illegal XML if something resembling an XML tag shows up in the error message or stack trace.

As an example, we get this Junit-XML which is clearly malformed:
```
<testcase classname="packages/server.faas.scrape/__tests__/scrape_basics.spec.ts" name="Extract title from title tag" time="3.3660000000">
            <failure message="expected &apos;&apos; to be &apos;English Title&apos; // Object.is equality" type="AssertionError">
AssertionError: expected '' to be 'English Title' // Object.is equality
 ❯ test_function packages/server.faas.scrape/__tests__/scrape_basics.spec.ts:26:30
 ❯ Server.<anonymous> packages/lib.shared.test/fixtures.ts:16:16
            </failure>
        </testcase>

```

The string `<anonymous>` should be escaped.

### Reproduction

See the following Stackblitz for an example of a test that produces malformed XML:

https://stackblitz.com/edit/vitest-dev-vitest-b4w9ta?file=test/basic.test.ts

### System Info

```shell
System:
    OS: Linux 5.10 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (4) x64 Intel(R) Xeon(R) CPU @ 2.20GHz
    Memory: 2.61 GB / 3.84 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 18.7.0 - /usr/bin/node
    Yarn: 3.1.0 - /usr/bin/yarn
    npm: 8.15.0 - /usr/bin/npm
  npmPackages:
    vitest: 0.21.0 => 0.21.0
```


### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vitest/src/node/reporters/junit.ts
+++ b/packages/vitest/src/node/reporters/junit.ts
@@ -112,7 +112,10 @@ export class JUnitReporter implements Reporter {
 
   async writeErrorDetails(error: ErrorWithDiff): Promise<void> {
     const errorName = error.name ?? error.nameStr ?? 'Unknown Error'
-    await this.baseLog(`${errorName}: ${error.message}`)
+    const errorDetails = `${errorName}: ${error.message}`
+
+    // Be sure to escape any XML in the error Details
+    await this.baseLog(escapeXML(errorDetails))
 
     const stack = parseStacktrace(error)
 

--- a/test/reporters/src/data-for-junit.ts
+++ b/test/reporters/src/data-for-junit.ts
@@ -1,0 +1,58 @@
+import { AssertionError } from 'assert'
+import type { File, Suite, Task } from 'vitest'
+
+function createSuiteHavingFailedTestWithXmlInError(): File[] {
+  const file: File = {
+    id: '1223128da3',
+    name: 'test/core/test/basic.test.ts',
+    type: 'suite',
+    mode: 'run',
+    filepath: '/vitest/test/core/test/basic.test.ts',
+    result: { state: 'fail', duration: 145.99284195899963 },
+    tasks: [],
+  }
+
+  const suite: Suite = {
+    id: '',
+    type: 'suite',
+    name: 'suite',
+    mode: 'run',
+    file,
+    result: { state: 'pass', duration: 1.90183687210083 },
+    tasks: [],
+  }
+
+  const errorWithXml = new AssertionError({
+    message: 'error message that has XML in it <tag>',
+  })
+
+  errorWithXml.stack = 'Error: error message that has XML in it <tag>\n'
+    + '    at /vitest/test/core/test/basic.test.ts:8:32\n'
+    + '    at etc....'
+
+  const tasks: Task[] = [
+    {
+      id: '123_0',
+      type: 'test',
+      name: 'test with xml in error',
+      mode: 'run',
+      suite,
+      fails: undefined,
+      file,
+      result: {
+        state: 'fail',
+        error: errorWithXml,
+        duration: 2.123123123,
+      },
+      context: null as any,
+    },
+  ]
+
+  file.tasks = [suite]
+  suite.tasks = tasks
+
+  return [file]
+}
+
+export { createSuiteHavingFailedTestWithXmlInError }
+

--- a/test/reporters/src/data.ts
+++ b/test/reporters/src/data.ts
@@ -176,58 +176,5 @@ suite.tasks = tasks
 
 const files = [file]
 
-function createSuiteHavingFailedTestWithXmlInError(): File[] {
-  const file: File = {
-    id: '1223128da3',
-    name: 'test/core/test/basic.test.ts',
-    type: 'suite',
-    mode: 'run',
-    filepath: '/vitest/test/core/test/basic.test.ts',
-    result: { state: 'fail', duration: 145.99284195899963 },
-    tasks: [],
-  }
-
-  const suite: Suite = {
-    id: '',
-    type: 'suite',
-    name: 'suite',
-    mode: 'run',
-    file,
-    result: { state: 'pass', duration: 1.90183687210083 },
-    tasks: [],
-  }
-
-  const errorWithXml = new AssertionError({
-    message: 'error message that has XML in it <tag>',
-  })
-
-  errorWithXml.stack = 'Error: error message that has XML in it <tag>\n'
-    + '    at /vitest/test/core/test/basic.test.ts:8:32\n'
-    + '    at etc....'
-
-  const tasks: Task[] = [
-    {
-      id: '123_0',
-      type: 'test',
-      name: 'test with xml in error',
-      mode: 'run',
-      suite,
-      fails: undefined,
-      file,
-      result: {
-        state: 'fail',
-        error: errorWithXml,
-        duration: 2.123123123,
-      },
-      context: null as any,
-    },
-  ]
-
-  file.tasks = [suite]
-  suite.tasks = tasks
-
-  return [file]
-}
-
-export { files, createSuiteHavingFailedTestWithXmlInError }
+export { files }
 

--- a/test/reporters/src/data.ts
+++ b/test/reporters/src/data.ts
@@ -174,4 +174,60 @@ const tasks: Task[] = [
 file.tasks = [suite]
 suite.tasks = tasks
 
-export const files = [file]
+const files = [file]
+
+function createSuiteHavingFailedTestWithXmlInError(): File[] {
+  const file: File = {
+    id: '1223128da3',
+    name: 'test/core/test/basic.test.ts',
+    type: 'suite',
+    mode: 'run',
+    filepath: '/vitest/test/core/test/basic.test.ts',
+    result: { state: 'fail', duration: 145.99284195899963 },
+    tasks: [],
+  }
+
+  const suite: Suite = {
+    id: '',
+    type: 'suite',
+    name: 'suite',
+    mode: 'run',
+    file,
+    result: { state: 'pass', duration: 1.90183687210083 },
+    tasks: [],
+  }
+
+  const errorWithXml = new AssertionError({
+    message: 'error message that has XML in it <tag>',
+  })
+
+  errorWithXml.stack = 'Error: error message that has XML in it <tag>\n'
+    + '    at /vitest/test/core/test/basic.test.ts:8:32\n'
+    + '    at etc....'
+
+  const tasks: Task[] = [
+    {
+      id: '123_0',
+      type: 'test',
+      name: 'test with xml in error',
+      mode: 'run',
+      suite,
+      fails: undefined,
+      file,
+      result: {
+        state: 'fail',
+        error: errorWithXml,
+        duration: 2.123123123,
+      },
+      context: null as any,
+    },
+  ]
+
+  file.tasks = [suite]
+  suite.tasks = tasks
+
+  return [file]
+}
+
+export { files, createSuiteHavingFailedTestWithXmlInError }
+

--- a/test/reporters/tests/reporters.spec.ts
+++ b/test/reporters/tests/reporters.spec.ts
@@ -6,7 +6,7 @@ import { JUnitReporter } from '../../../packages/vitest/src/node/reporters/junit
 import { TapReporter } from '../../../packages/vitest/src/node/reporters/tap'
 import { TapFlatReporter } from '../../../packages/vitest/src/node/reporters/tap-flat'
 import { getContext } from '../src/context'
-import { files } from '../src/data'
+import { createSuiteHavingFailedTestWithXmlInError, files } from '../src/data'
 
 afterEach(() => {
   vi.useRealTimers()
@@ -93,6 +93,35 @@ test('JUnit reporter with outputFile', async () => {
   // Act
   await reporter.onInit(context.vitest)
   await reporter.onFinished(files)
+
+  // Assert
+  expect(normalizeCwd(context.output)).toMatchSnapshot()
+  expect(existsSync(outputFile)).toBe(true)
+  expect(readFileSync(outputFile, 'utf8')).toMatchSnapshot()
+
+  // Cleanup
+  rmSync(outputFile)
+})
+
+test('JUnit reporter with outputFile with XML in error details', async () => {
+  // Arrange
+  const reporter = new JUnitReporter()
+  const outputFile = resolve('report.xml')
+  const context = getContext()
+  context.vitest.config.outputFile = outputFile
+
+  vi.mock('os', () => ({
+    hostname: () => 'hostname',
+  }))
+
+  vi.setSystemTime(1642587001759)
+
+  // setup test with failed test with xml
+  const filesWithTestHavingXmlInError = createSuiteHavingFailedTestWithXmlInError()
+
+  // Act
+  await reporter.onInit(context.vitest)
+  await reporter.onFinished(filesWithTestHavingXmlInError)
 
   // Assert
   expect(normalizeCwd(context.output)).toMatchSnapshot()

--- a/test/reporters/tests/reporters.spec.ts
+++ b/test/reporters/tests/reporters.spec.ts
@@ -103,10 +103,10 @@ test('JUnit reporter with outputFile', async () => {
   rmSync(outputFile)
 })
 
-test('JUnit reporter with outputFile with XML in error details', async () => {
+test('JUnit reporter with outputFile with XML in error message', async () => {
   // Arrange
   const reporter = new JUnitReporter()
-  const outputFile = resolve('report.xml')
+  const outputFile = resolve('report_escape_msg_xml.xml')
   const context = getContext()
   context.vitest.config.outputFile = outputFile
 

--- a/test/reporters/tests/reporters.spec.ts
+++ b/test/reporters/tests/reporters.spec.ts
@@ -6,7 +6,8 @@ import { JUnitReporter } from '../../../packages/vitest/src/node/reporters/junit
 import { TapReporter } from '../../../packages/vitest/src/node/reporters/tap'
 import { TapFlatReporter } from '../../../packages/vitest/src/node/reporters/tap-flat'
 import { getContext } from '../src/context'
-import { createSuiteHavingFailedTestWithXmlInError, files } from '../src/data'
+import { files } from '../src/data'
+import { createSuiteHavingFailedTestWithXmlInError } from '../src/data-for-junit'
 
 afterEach(() => {
   vi.useRealTimers()
@@ -116,7 +117,7 @@ test('JUnit reporter with outputFile with XML in error message', async () => {
 
   vi.setSystemTime(1642587001759)
 
-  // setup test with failed test with xml
+  // setup suite with failed test with xml
   const filesWithTestHavingXmlInError = createSuiteHavingFailedTestWithXmlInError()
 
   // Act

